### PR TITLE
Remove 'simplification' and other associated tests #1362

### LINKED ISSUE #1362: Remove `Simplification` class and associated tests
Following #1301, the `Simplification` class is no longer required (see https://github.com/pybamm-team/PyBaMM/pull/1359#issuecomment-773442388)

The following can be removed:
- everything in `simplify.py` **except** `simplify_in_constant` (which could be moved to `symbol.py`)
- all simplifications in the `BaseSolver` class
- the `use_simplify` attribute of a model
- all `def simplify` function methods (add a clear error message to the one in `Symbol`)
- anywhere `.simplify()` appears
- testing simplifications in the `OptimisationsTest`
- all other associated tests and docs

I don't think anyone is using this so we are good to just delete instead of deprecating, imo

This is a bit of a stretch for a "good first issue" but should be doable for someone new unless the tests break in an unexpected way

### PR DESCRIPTION
**Pull request by asinghgaba**
# Description

Removed `Simplification` class and relevant tests associated with it. Also, moved `simplify_if_constant()` to `symbol.py`.
Fixes #1362 
Completed all the tasks specified in #1362 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas (**not required**)
- [ ] Tests added that prove fix is effective or that feature works (**not required**)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
 ## Bug fixes
 
+-   Removed class `Simplification` ([#1369](https://github.com/pybamm-team/PyBaMM/pull/1369))
```
Comment (tinosulzer) [inline CHANGELOG.md:32]:
```suggestion
-   Removed `Simplification` class and `.simplify()` function ([#1369](https://github.com/pybamm-team/PyBaMM/pull/1369))
```
and this should go under "breaking changes" instead

File: tests/integration/test_models/test_full_battery_models/test_lead_acid/test_composite.py [hunk 2]
```diff
@@ -28,22 +28,20 @@ def test_optimisations(self):
         optimtest = tests.OptimisationsTest(model)
 
         original = optimtest.evaluate_model()
-        simplified = optimtest.evaluate_model(simplify=True)
         using_known_evals = optimtest.evaluate_model(use_known_evals=True)
-        simp_and_known = optimtest.evaluate_model(simplify=True, use_known_evals=True)
-        simp_and_python = optimtest.evaluate_model(simplify=True, to_python=True)
-        np.testing.assert_array_almost_equal(original, simplified)
+        simp_and_known = optimtest.evaluate_model(use_known_evals=True)
```
Comment (tinosulzer) [inline tests/integration/test_models/test_full_battery_models/test_lead_acid/test_composite.py:32]: this is now a duplicate of the above line (same in all other tests)

File: tests/unit/test_expression_tree/test_concatenations.py [hunk 3]
```diff
@@ -312,25 +312,6 @@ def test_domain_error(self):
...
-        )
-        self.assertEqual(
-            pybamm.NumpyConcatenation(a, pybamm.NumpyConcatenation(b, c)).simplify().id,
-            pybamm.NumpyConcatenation(a, b, c).id,
-        )
-        # check it works when calling numpy_concatenation
-        self.assertEqual(
-            pybamm.numpy_concatenation(pybamm.numpy_concatenation(a, b), c).id,
-            pybamm.NumpyConcatenation(a, b, c).id,
-        )
```
Comment (tinosulzer) [inline tests/unit/test_expression_tree/test_concatenations.py:332]: can we keep this part of the test?

File: pybamm/expression_tree/symbol.py [hunk 4]
```diff
@@ -788,7 +824,7 @@ def has_symbol_of_classes(self, symbol_classes):
 
     def simplify(self, simplified_symbols=None, clear_domains=True):
         """ Simplify the expression tree. See :class:`pybamm.Simplification`. """
-        return pybamm.Simplification(simplified_symbols).simplify(self, clear_domains)
+        raise pybamm.ModelError(".simplify() is deprecated as it now has no effect")
```
Comment (tinosulzer) [inline pybamm/expression_tree/symbol.py:827]: need to add a test for this line

**Review by tinosulzer — CHANGES_REQUESTED** [review]
Thanks @asinghgaba! Just a couple of small things to change. Let's also wait to see if the coverage is still ok, I think it should be

File: CHANGELOG.md (hunk 1 above)
Comment (asinghgaba) [reply CHANGELOG.md:32]: Okay 👍

File: tests/integration/test_models/test_full_battery_models/test_lead_acid/test_composite.py (hunk 2 above)
Comment (asinghgaba) [reply tests/integration/test_models/test_full_battery_models/test_lead_acid/test_composite.py:32]: I'm sorry I didn't realise that. I'll change that.

File: tests/unit/test_expression_tree/test_concatenations.py (hunk 3 above)
Comment (asinghgaba) [reply tests/unit/test_expression_tree/test_concatenations.py:332]: Ah yes I checked we can keep it, I'll add it back.

File: pybamm/expression_tree/symbol.py (hunk 4 above)
Comment (asinghgaba) [reply pybamm/expression_tree/symbol.py:827]: I'm sorry but what do you mean by test for this line? (Apologies if this is a foolish question but I'm new to open source programming)

Comment (tinosulzer) [reply pybamm/expression_tree/symbol.py:827]: We can test that errors are raised when expected, for example like in this line
https://github.com/pybamm-team/PyBaMM/blob/develop/tests/unit/test_citations.py#L21

Comment (asinghgaba) [reply pybamm/expression_tree/symbol.py]: Thank you @tinosulzer , I have now added the test.

**Review by tinosulzer — CHANGES_REQUESTED** [review]
It turns out there are a few more things to delete (from the coverage report):
- all the `_unary_simplify` functions
- all the `_binary_simplify` functions
- all the `_concatenation_simplify` functions

Then it should be done! There are some other coverage changes but each one requires different action so we can take care of later

**Review by tinosulzer — APPROVED** [review]
all good, merging

**Comment by tinosulzer** [comment]
@all-contributors add @asinghgaba for code

**Pull request merged**